### PR TITLE
Upgrade to groovy 2.5

### PR DESCRIPTION
**Pull request by lptr**
Changes:

- Upgraded to Groovy 2.5.2.

    This required building our own `groovy-all.jar` (see https://github.com/gradle/gradle-groovy-all), because Groovy 2.5 stopped shipping the uber-jar and replaced it with a BOM. I tried to use the separate JARs, but it poses a number of issues, most importantly that the increased length of the classpath breaks our Windows tests. So we decided to go with repackaging Groovy. In the future it would be nice to remove the need for the repackaging, and maybe even remove some of the Groovy libs. I mean, c'mon, we ship classes from `groovy-swing`!

- Downgraded `opentest4j` to work around https://github.com/ota4j-team/opentest4j/issues/49, introduced by an upgrade in TestNG.
- Upgraded compiler memory in IntelliJ from 2 GB to 3 GB because Groovy compilation with 2.5 seemed to slow down a lot, and produce OOMEs occasionally. With the increased amount of memory Groovy compilation is actually quite snappy.
- Replaced calls to `DateGroovyMethods` extension methods. This led to the realization that some of our timeouts mistakenly use days instead of milliseconds. We should fix these separately, added `TODO`s.
- Bumped Spock to the Groovy 2.5 variant and fixed a number of issues with our tests. Some of these issues are valid ones, others seem to indicate bugs in either the Groovy compiler or Spock. Reported the issue to Spock: https://github.com/spockframework/spock/issues/880

    The `Specification` class added by @ldaley to fix type inference in IntelliJ doesn't work anymore unfortunately. I left the code commented out so it's easy to fix it if someone wants to.

- Groovy 2.5 adds local variable type information that needs to be remapped in our script remapper, so I added that.
- Groovy 2.5 uses the `ExtensionModule` mechanism slightly differently which necessitated the exposure of some more packages and resources from `gradleApi()` (see changes in `DefaultGradleApiSpecProvider.java`)
- Replaced references to types in `com.sun.jdi` in `JDWPUtil` with reflection to work around an IntelliJ bug we hit: https://youtrack.jetbrains.com/oauth?state=%2Fissue%2FIDEA-198715
- There is some weirdness wrt Tooling API progress tests where the order of events seems to have changed somewhat. See changes in `ToolingApiBuildExecutionCrossVersionSpec.groovy` and `ToolingApiModelCrossVersionSpec.groovy`.
- Stopped using Groovy JSON in `:buildSrc:binaryCompatibilityCheck` because of a classloading issue with Groovy's `FastStringUtils`.

Supersedes #6324.

### CODE REVIEW

File: subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/DependencyHandlerApiResolveIntegrationTest.groovy [hunk 1]
```diff
@@ -140,7 +140,8 @@ class DependencyHandlerApiResolveIntegrationTest extends AbstractIntegrationSpec
         then:
         def gradleVersion = GradleVersion.current().version
         def gradleBaseVersion = GradleVersion.current().baseVersion.version
-        def groovyVersion = GroovySystem.version
+        // The suffix comes from https://github.com/gradle/gradle-groovy-all
+        def groovyVersion = GroovySystem.version + "-0.3"
```
[the remark this answers is not in the document]
Comment (lptr) [reply subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/DependencyHandlerApiResolveIntegrationTest.groovy]: Good point!

[the remark this answers is not in the document]
Comment (lptr) [reply subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/DependencyHandlerApiResolveIntegrationTest.groovy]: I've ended up adding a version property in our build of the `groovy-all.jar`.

**Comment by lptr** [comment]
@marcphilipp Should be fixed now, PTAL.

**Comment by lptr** [comment]
I've added some notes to the PR to explain the changes.

**Comment by lptr** [comment]
OK, looks like the Gradleception build is fixed, and thus we have no failing functional tests anymore. There are some performance test failures for some reason here that I don't yet understand. I'd vote for merging this and fixing them on `master` so that we can unblock other work waiting for this to be merged. WDYT @oehme?

**Comment by lptr** [comment]
Failing performance tests: https://builds.gradle.org/viewLog.html?buildId=15812193&problemId=19827

**Comment by oehme** [comment]
The same tests are failing on master: https://builds.gradle.org/viewLog.html?buildId=15810181&buildTypeId=Gradle_Check_PerformanceTestCoordinator

I'm fine with merging this, but after that we should put up a merge ban, since performance on master has regressed in many scenarios and it hasn't been investigated yet.

File: buildSrc/subprojects/configuration/src/main/kotlin/org/gradle/gradlebuild/dependencies/DependenciesMetadataRulesPlugin.kt [hunk 2]
```diff
@@ -227,27 +229,13 @@ fun ComponentMetadataHandler.withLibraryDependencies(module: String, kClass: KCl
...
-open class DowngradeTestNGRule : ComponentMetadataRule {
-    override fun execute(context: ComponentMetadataContext) {
-        context.details.allVariants {
-            withDependencyConstraints {
-                filter { it.group == "org.testng" }.forEach {
-                    it.version { prefer("6.3.1") }
-                    it.because("6.3.1 is required by Gradle and part of the distribution")
+            withDependencies {
+                filter { it.group == "org.opentest4j" }.forEach {
+                    it.version { strictly("1.0.0") }
```
Comment (wolfs) [inline buildSrc/subprojects/configuration/src/main/kotlin/org/gradle/gradlebuild/dependencies/DependenciesMetadataRulesPlugin.kt]: @marcphilipp Do we now have an `opentest4j` version where this problem is fixed?

Comment (marcphilipp) [reply buildSrc/subprojects/configuration/src/main/kotlin/org/gradle/gradlebuild/dependencies/DependenciesMetadataRulesPlugin.kt]: Yes, and I'll update to it in #6728.

File: subprojects/tooling-api/src/crossVersionTest/groovy/org/gradle/integtests/tooling/m5/ToolingApiBuildExecutionCrossVersionSpec.groovy [hunk 3]
```diff
@@ -81,7 +81,8 @@ System.err.println 'this is stderr'
 
         then:
         progress.size() >= 2
-        progress.pop() == ''
+        // TODO:lptr I have no idea why this changed from pop() to removeLast()
```
[the remark this answers is not in the document]
Comment (lptr) [reply subprojects/tooling-api/src/crossVersionTest/groovy/org/gradle/integtests/tooling/m5/ToolingApiBuildExecutionCrossVersionSpec.groovy]: Oh, nice. Thanks. Wow.

File: subprojects/internal-testing/src/main/groovy/org/gradle/testing/internal/util/Specification.groovy [hunk 4]
```diff
@@ -17,35 +17,33 @@
...
 import groovy.transform.CompileStatic
-import groovy.transform.stc.ClosureParams
-import groovy.transform.stc.SecondParam
-
 /**
  * This fixes Spock's Mock() and Stub() methods to be understandable by IDEA.
  * By using this subclass, IDEA knows the closure delegate type when using Mock() or Stub().
  * This means that you get editing assistance in said blocks WRT the type being mocked/stubbed.
  */
+// TODO:lptr Either remove this class, or fix the two methods with Groovy 2.5
```
[the remark this answers is not in the document]
Comment (lptr) [reply subprojects/internal-testing/src/main/groovy/org/gradle/testing/internal/util/Specification.groovy]: I'd love to fix this better. I created https://github.com/gradle/gradle-private/issues/1509, let's see if Luke (or someone) has the time.

File: subprojects/internal-testing/src/main/groovy/org/gradle/test/fixtures/ConcurrentTestUtil.groovy [hunk 5]
```diff
@@ -385,8 +385,9 @@ abstract class AbstractAction implements LongRunningAction {
     }
 
     void completesWithin(long maxWaitValue, TimeUnit maxWaitUnits) {
-        Date expiry = new Date(System.currentTimeMillis() + maxWaitUnits.toMillis(maxWaitValue))
-        completesBefore(expiry + 500)
+        long expiry = System.currentTimeMillis() + maxWaitUnits.toMillis(maxWaitValue)
+        // TODO I'm pretty sure we should not add 500 _days_ here, but changing that kills some Play integration tests
+        completesBefore(new Date(expiry + TimeUnit.DAYS.toMillis(500)))
```
[the remark this answers is not in the document]
Comment (lptr) [reply subprojects/internal-testing/src/main/groovy/org/gradle/test/fixtures/ConcurrentTestUtil.groovy]: I spent some time on this, but failed to run the tests locally. I'd say if this is important, we should fix it separately. Please create an issue if you feel strongly about this.

File: buildSrc/subprojects/binary-compatibility/src/main/groovy/org/gradle/binarycompatibility/rules/AbstractGradleViolationRule.groovy [hunk 6]
```diff
@@ -170,13 +170,22 @@ abstract class AbstractGradleViolationRule extends AbstractContextAwareViolation
...
         )
         return violation
     }
 
+    String prettyPrintJson(def acceptanceJson) {
+        def stringWriter = new StringWriter()
+        def writer = new JsonWriter(stringWriter)
+        writer.setIndent("    ")
+        new Gson().toJson(acceptanceJson, AcceptedApiChange, writer)
+        writer.close()
```
[the remark this answers is not in the document]
Comment (lptr) [reply buildSrc/subprojects/binary-compatibility/src/main/groovy/org/gradle/binarycompatibility/rules/AbstractGradleViolationRule.groovy]: This is actually a bit nicer indeed, I'll change it:

```groovy
def stringWriter = new StringWriter()
new JsonWriter(stringWriter).withCloseable { writer ->
    writer.setIndent("    ")
    new Gson().toJson(acceptanceJson, AcceptedApiChange, writer)
}
return stringWriter.toString()
```

File: gradle/dependency-management/capabilities.json [hunk 7]
```diff
@@ -44,6 +44,14 @@
         ],
         "selected": "org.slf4j:log4j-over-slf4j"
     },
+    {
+        "name": "groovy",
+        "providedBy": [
+            "org.codehaus.groovy:groovy",
```
[the remark this answers is not in the document]
Comment (wolfs) [reply gradle/dependency-management/capabilities.json]: Shall we then add a conflict in some way between our groovy-all and any of the other groovy libraries? That would make sure that the build fails should one of them be on the classpath.

**Comment by lptr** [comment]
@wolfs PTAL.

**Comment by lptr** [comment]
Changes are here: https://github.com/gradle/gradle/pull/6799

**Pull request merged**